### PR TITLE
[pipeline] integration

### PR DESCRIPTION
**Pull request by zekun000**
## Description
This pr integrates the pipeline framework from https://github.com/aptos-labs/aptos-core/pull/15145 to existing code, and use a local config to gate it.

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: consensus/src/pipeline/buffer_manager.rs [hunk 1]
```diff
@@ -557,6 +558,27 @@ impl BufferManager {
...
+                        commit_ledger_fut,
+                        ..
+                    } = fut;
+                    let _ = join4(
+                        execute_fut,
+                        ledger_update_fut,
+                        pre_commit_fut,
+                        commit_ledger_fut,
+                    )
+                    .await;
```
Comment (msmouse) [inline consensus/src/pipeline/buffer_manager.rs:578]: deserves `PipelineFuts::join(self)`?

File: consensus/src/pipeline/buffer_manager.rs [hunk 2]
```diff
@@ -557,6 +558,27 @@ impl BufferManager {
...
+                        ledger_update_fut,
+                        pre_commit_fut,
+                        commit_ledger_fut,
+                        ..
+                    } = fut;
+                    let _ = join4(
+                        execute_fut,
+                        ledger_update_fut,
+                        pre_commit_fut,
+                        commit_ledger_fut,
```
Comment (msmouse) [inline consensus/src/pipeline/buffer_manager.rs:576]: thought you just need to join the "last" fut?

File: consensus/src/pipeline/execution_schedule_phase.rs [hunk 3]
```diff
@@ -75,37 +75,60 @@ impl StatelessPipeline for ExecutionSchedulePhase {
...
+            async move {
+                for b in ordered_blocks.iter_mut() {
+                    let (compute_result, _) = b
+                        .pipeline_fut()
+                        .unwrap()
+                        .ledger_update_fut
+                        .clone()
+                        .await
+                        .map_err(|_| ExecutorError::InternalError {
+                            error: "lalala".to_string(),
```
Comment (msmouse) [inline consensus/src/pipeline/execution_schedule_phase.rs:94]: well done, boss, well done

File: consensus/src/pipeline/execution_schedule_phase.rs [hunk 4]
```diff
@@ -61,12 +61,12 @@ impl StatelessPipeline for ExecutionSchedulePhase {
...
         let ExecutionRequest {
-            ordered_blocks,
+            mut ordered_blocks,
             lifetime_guard,
         } = req;
 
-        let block_id = match ordered_blocks.last() {
-            Some(block) => block.id(),
+        let (block_id, pipeline_enabled) = match ordered_blocks.last() {
+            Some(block) => (block.id(), block.pipeline_enabled()),
```
Comment (msmouse) [inline consensus/src/pipeline/execution_schedule_phase.rs:69]: why is the flag carried on the block again?

File: consensus/src/pipeline/execution_schedule_phase.rs [hunk 5]
```diff
@@ -75,37 +75,60 @@ impl StatelessPipeline for ExecutionSchedulePhase {
...
+                    let (compute_result, _) = b
+                        .pipeline_fut()
+                        .unwrap()
+                        .ledger_update_fut
+                        .clone()
+                        .await
+                        .map_err(|_| ExecutorError::InternalError {
+                            error: "lalala".to_string(),
+                        })?;
+                    b.set_compute_result(compute_result);
```
Comment (msmouse) [inline consensus/src/pipeline/execution_schedule_phase.rs:96]: can be `b.wait_for_compute_result().await`

File: consensus/src/pipeline/persisting_phase.rs [hunk 6]
```diff
@@ -67,11 +67,22 @@ impl StatelessPipeline for PersistingPhase {
...
-        self.persisting_handle
-            .commit(&blocks, commit_ledger_info, callback)
-            .await
-            .map(|_| round)
+        if blocks.last().unwrap().pipeline_enabled() {
+            for b in &blocks {
+                if let Some(tx) = b.pipeline_tx().lock().as_mut() {
+                    let _ = tx.commit_proof_tx.send(commit_ledger_info.clone());
+                }
+                let _ = b.pipeline_fut().unwrap().commit_ledger_fut.clone().await;
```
Comment (msmouse) [inline consensus/src/pipeline/persisting_phase.rs:76]: b.commit().await;

File: consensus/src/pipeline/execution_schedule_phase.rs (hunk 4 above)
Comment (zekun000) [reply consensus/src/pipeline/execution_schedule_phase.rs:69]: to determine whether it uses the new pipeline or not?

File: consensus/src/consensus_observer/observer/active_state.rs [hunk 7]
```diff
@@ -108,12 +107,31 @@ impl ActiveObserverState {
...
+                pending_ordered_blocks,
+                block_payload_store,
+                root,
+                // blocks,
+                ledger_info,
+            );
+        })
+    }
+
+    pub fn create_commit_callback_v2(
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/consensus_observer/observer/active_state.rs:121]: the old callback signature requires a (blocks, ledger) closures

File: consensus/src/test_utils/mock_execution_client.rs [hunk 8]
```diff
@@ -186,4 +187,8 @@ impl TExecutionClient for MockExecutionClient {
     }
 
     async fn end_epoch(&self) {}
+
+    fn pipeline_builder(&self, _signer: Arc<ValidatorSigner>) -> PipelineBuilder {
+        todo!()
```
Comment (msmouse) [inline consensus/src/test_utils/mock_execution_client.rs:192]: `todo!` or `unimplemented!` / `unreachable!`?

File: consensus/src/twins/twins_node.rs [hunk 9]
```diff
@@ -279,6 +279,7 @@ impl SMRNode {
             config.base.waypoint = WaypointConfig::FromConfig(waypoint);
             // Disable timeout in twins test to avoid flakiness
             config.consensus.round_initial_timeout_ms = 2_000_000;
+            config.consensus.enable_pipeline = false;
```
Comment (msmouse) [inline consensus/src/twins/twins_node.rs:282]: 🤔

File: consensus/src/consensus_observer/observer/consensus_observer.rs [hunk 10]
```diff
@@ -337,13 +347,27 @@ impl ConsensusObserver {
...
-            last_ordered_block
+            last_ordered_block.block_info()
         } else {
             // Return the root ledger info
             self.active_observer_state.root().commit_info().clone()
         }
     }
 
+    /// Returns the last block's pipeline futures, should only be called when pipeline is enabled
+    fn get_last_pipeline_fut(&self) -> PipelineFutures {
```
Comment (msmouse) [inline consensus/src/consensus_observer/observer/consensus_observer.rs:358]: futs

File: consensus/src/consensus_observer/observer/consensus_observer.rs [hunk 11]
```diff
@@ -755,6 +779,19 @@ impl ConsensusObserver {
                 metrics::ORDERED_BLOCK_LABEL,
             );
 
+            if self.active_observer_state.pipeline_enabled() {
```
Comment (JoshLind) [inline consensus/src/consensus_observer/observer/consensus_observer.rs:782]: nit: `self.pipeline_enabled()?`

File: consensus/src/consensus_observer/observer/consensus_observer.rs [hunk 12]
```diff
@@ -973,6 +1010,9 @@ impl ConsensusObserver {
                 0,
             )
             .await;
+        if self.active_observer_state.pipeline_enabled() {
```
Comment (JoshLind) [inline consensus/src/consensus_observer/observer/consensus_observer.rs:1013]: nit: `self.pipeline_enabled()?`

**Review by JoshLind — COMMENTED** [review]
CO changes LGTM 😄 Thanks boss!

**Review by JoshLind — APPROVED** [review]
(Will add my stamp and let @msmouse do the real stamp 😄)

File: consensus/src/twins/twins_node.rs (hunk 9 above)
Comment (danielxiangzl) [reply consensus/src/twins/twins_node.rs:282]: Wen remove this folder

**Review by msmouse — APPROVED** [review]
(Will add my stamp and the real stamp is in boss'es heart 💟 )

File: consensus/src/pipeline/pipeline_builder.rs [hunk 13]
```diff
@@ -192,8 +192,9 @@ impl PipelineBuilder {
         commit_proof: LedgerInfoWithSignatures,
     ) -> PipelineFutures {
         let prepare_fut = spawn_ready_fut(Arc::new(vec![]));
-        let execute_fut = spawn_ready_fut(());
-        let ledger_update_fut = spawn_ready_fut((compute_result.clone(), None));
+        let execute_fut = spawn_ready_fut(Duration::from_millis(0));
```
Comment (danielxiangzl) [inline consensus/src/pipeline/pipeline_builder.rs:195]: Why not put timer in the tracker struct?

**Review by danielxiangzl — APPROVED** [review]

File: consensus/src/pipeline/pipeline_builder.rs (hunk 13 above)
Comment (zekun000) [reply consensus/src/pipeline/pipeline_builder.rs:195]: I need to propagate the duration to the outside, it's used for backpressure, but I'll add some internal counters too

File: testsuite/smoke-test/src/smoke_test_environment.rs [hunk 14]
```diff
@@ -17,7 +17,7 @@ use rand::rngs::OsRng;
 use std::{num::NonZeroUsize, sync::Arc};
 use tokio::task::JoinHandle;
 
-const SWARM_BUILD_NUM_RETRIES: u8 = 3;
+const SWARM_BUILD_NUM_RETRIES: u8 = 0;
```
Comment (ibalajiarun) [inline testsuite/smoke-test/src/smoke_test_environment.rs:20]: undo?

File: consensus/src/pipeline/persisting_phase.rs [hunk 15]
```diff
@@ -67,11 +67,26 @@ impl StatelessPipeline for PersistingPhase {
...
-            .await
-            .map(|_| round)
+        if blocks
+            .last()
+            .expect("Blocks can't be empty")
+            .pipeline_enabled()
+        {
+            for b in &blocks {
+                if let Some(tx) = b.pipeline_tx().lock().as_mut() {
+                    let _ = tx.commit_proof_tx.send(commit_ledger_info.clone());
```
Comment (ibalajiarun) [inline consensus/src/pipeline/persisting_phase.rs:78]: same as previous: probably log error.

**Review by ibalajiarun — APPROVED** [review]

**Pull request merged**